File: backend/app/core/pathways.py
"""
Centralized Academic Pathway definitions and canonicalization.

Defines canonical DB codes, legacy alias resolution, and compatibility rules.
Kept strictly synchronized with frontend/src/lib/pathways.ts.
"""

from __future__ import annotations

# Canonical DB codes matching academic_pathway.code
CANONICAL_MEDICAL = "medical-courses"
CANONICAL_ENGINEERING = "engineering-math"
PATHWAY_BOTH = "both"
PATHWAY_GENERAL = "general"

# Exact, evidence-based alias mapping: legacy codes -> canonical DB codes
PATHWAY_ALIAS_MAP: dict[str, str] = {
    # Canonical DB codes
    "medical-courses": CANONICAL_MEDICAL,
    "engineering-math": CANONICAL_ENGINEERING,
    "general": PATHWAY_GENERAL,
    "both": PATHWAY_BOTH,
    # In-use legacy aliases
    "stem_medical": CANONICAL_MEDICAL,
    "stem_engineering": CANONICAL_ENGINEERING,
}


def canonicalize_pathway(code: str | None) -> str:
    """Normalize a pathway code or alias to its canonical value."""
    if not code:
        return PATHWAY_GENERAL
    normalized = str(code).strip().lower()
    return PATHWAY_ALIAS_MAP.get(normalized, normalized)


def are_pathways_compatible(offering_pathway: str | None, class_pathway: str | None) -> bool:
    """
    Check if a subject offering's pathway is compatible with a class's pathway.
    
    Compatible if:
    - Offering is shared across pathways ('both')
    - Both resolve to the same canonical pathway
    - Both resolve to 'general'
    """
    canon_offering = canonicalize_pathway(offering_pathway)
    canon_class = canonicalize_pathway(class_pathway)

    if canon_offering == PATHWAY_BOTH:
        return True
    if canon_offering == canon_class:
        return True
    if canon_offering == PATHWAY_GENERAL and canon_class == PATHWAY_GENERAL:
        return True

    return False

File: frontend/src/lib/pathways.ts
/**
 * Centralized Academic Pathway definitions and canonicalization.
 *
 * Defines canonical DB codes, legacy alias resolution, and compatibility rules.
 * Kept strictly synchronized with backend/app/core/pathways.py.
 */

export const CANONICAL_PATHWAYS = {
  MEDICAL: "medical-courses",
  ENGINEERING: "engineering-math",
  BOTH: "both",
  GENERAL: "general",
} as const;

export type CanonicalPathway =
  | (typeof CANONICAL_PATHWAYS)[keyof typeof CANONICAL_PATHWAYS]
  | string;

export const PATHWAY_ALIAS_MAP: Record<string, string> = {
  // Canonical DB codes
  "medical-courses": CANONICAL_PATHWAYS.MEDICAL,
  "engineering-math": CANONICAL_PATHWAYS.ENGINEERING,
  "general": CANONICAL_PATHWAYS.GENERAL,
  "both": CANONICAL_PATHWAYS.BOTH,
  // In-use legacy aliases
  "stem_medical": CANONICAL_PATHWAYS.MEDICAL,
  "stem_engineering": CANONICAL_PATHWAYS.ENGINEERING,
};

/**
 * Normalizes any pathway code or alias to its canonical value.
 */
export function canonicalizePathway(raw?: string | null): string {
  if (!raw) return CANONICAL_PATHWAYS.GENERAL;
  const normalized = String(raw).trim().toLowerCase();
  return PATHWAY_ALIAS_MAP[normalized] ?? normalized;
}

/**
 * Checks if a subject offering's pathway is compatible with a class's pathway.
 */
export function isOfferingCompatibleWithClass(
  offeringPathway?: string | null,
  classPathway?: string | null
): boolean {
  const canonOffering = canonicalizePathway(offeringPathway);
  const canonClass = canonicalizePathway(classPathway);

  if (canonOffering === CANONICAL_PATHWAYS.BOTH) {
    return true;
  }
  if (canonOffering === canonClass) {
    return true;
  }
  if (canonOffering === CANONICAL_PATHWAYS.GENERAL && canonClass === CANONICAL_PATHWAYS.GENERAL) {
    return true;
  }

  return false;
}

/**
 * Returns human-readable label for a pathway, using the loaded pathways list from DB if available.
 */
export function getPathwayDisplayName(
  code?: string | null,
  pathwaysList?: Array<{ code: string; name: string }>
): string {
  const canon = canonicalizePathway(code);
  if (canon === CANONICAL_PATHWAYS.BOTH) return "Shared / All Pathways";
  if (canon === CANONICAL_PATHWAYS.GENERAL) return "General";

  const found = pathwaysList?.find((p) => canonicalizePathway(p.code) === canon);
  if (found?.name) return found.name;

  if (canon === CANONICAL_PATHWAYS.MEDICAL) return "Medical Courses and Sciences Related";
  if (canon === CANONICAL_PATHWAYS.ENGINEERING) return "Engineering and Mathematics Related";

  return canon;
}
